### PR TITLE
modified transforms.py to accept list of PIL images 

### PR DESCRIPTION
**Pull request by quantummole**
if input is a list of images then output is a list of images else old behaviour is retained
if params need to be computed for every img, then they are computed based on the fist img of the list
this was done to apply the same random transform on each img for example - image segmentation

### CODE REVIEW

**Comment by fmassa** [comment]
Hi,

Thanks for the PR!

I'm not sure that this is the way we would want those problems to be handled. It indeed adds a strong assumption that everything in the list if a PIL Image.
This is the case for Image Segmentation, but still breaks some assumptions even in this case.
For example, we don't want to apply bilinear interpolation for the ground-truth segmentation masks, so this should be handled differently and that's not covered in this PR. Same thing for rotations, scalings, etc.

For this reason, I think that we really want to keep the class interface as simple as possible, and let the user leverage the functional interface to perform their fine-grained transformations themselves.

What do you think?

**Comment by quantummole** [comment]
You do have a point. Then is it possible for us to have a uniform interface like get_params  for transforms like random_grayscale etc which would return a random bool that we can use to chain methods inside the dataset.

**Comment by fmassa** [comment]
Oh, yes, definitely!
`RandomGrayscale` should have a `get_params` in there, it was an oversight that it didn't have it.
Ideally, all transforms that have random parameters should have them via a `get_params` staticmethod.

Could you send a PR fixing this in `RandomGrayscale`?

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -503,6 +515,16 @@ class RandomVerticalFlip(object):
     def __init__(self, p=0.5):
         self.p = p
 
+    def get_params(p):
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: This should be a staticmethod

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -485,7 +496,8 @@ def __call__(self, img):
         Returns:
             PIL Image: Randomly flipped image.
         """
-        if random.random() < self.p:
+        to_flip = self.get_params(self.p)
+        if to_flip :
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: nit: space after `:` is not compliant with our linter

File: torchvision/transforms/transforms.py [hunk 3]
```diff
@@ -511,7 +533,8 @@ def __call__(self, img):
...
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: nit: space

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -1037,6 +1060,7 @@ class Grayscale(object):
     def __init__(self, num_output_channels=1):
         self.num_output_channels = num_output_channels
 
+
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: nit: remove the newline

File: torchvision/transforms/transforms.py [hunk 5]
```diff
@@ -1068,6 +1092,16 @@ class RandomGrayscale(object):
     def __init__(self, p=0.1):
         self.p = p
 
+    def get_params(p):
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: staticmethod

File: torchvision/transforms/transforms.py [hunk 6]
```diff
@@ -1077,7 +1111,8 @@ def __call__(self, img):
             PIL Image: Randomly grayscaled image.
         """
         num_output_channels = 1 if img.mode == 'L' else 3
-        if random.random() < self.p:
+        to_convert = self.get_params(self.p)
+        if to_convert :
```
Comment (fmassa) [inline torchvision/transforms/transforms.py]: nit: space after `:`

**Comment by fmassa** [comment]
Actually, now I remember why we didn't have `get_params` for those functions: it was because they were fairly trivial, and generally simpler to implement in the code itself.
I'm not against adding it here, but I'd like to know how you are planning to write your code to support the semantic segmentation use-case

**Comment by fmassa** [comment]
The problem is that `get_params` is a staticmethod, so it's not that easy to abstract that away.
In those cases, I think it's just way simpler to write your own transform, using purely the functional interface

```python
class MyTransform(object):
    def __init__(self, **kwargs):
        # store params here

    def __call__(self, img, masks, etc):
        if random.rand() > self.prob1:
               # do my magic
        return img, masks, etc
```

**Pull request closed without merging**